### PR TITLE
Filter Manager: Local Reply no longer filter reentrant

### PR DESCRIPTION
**Pull request by KBaichoo**
Commit Message: Filter Manager: Local Reply no longer filter reentrant
Additional Description: This change should not affect the dataplane in a user facing way. It removes an edge case where a filter can be invoked by both the decoder filter chain and the encoder filter chain at the same time.
Risk Level: medium (added reloadable flag)
Testing: parameterized tests 
Docs Changes: n/a
Release Notes: yes
Platform Specific Features: n/a
Runtime guard: yes

### CODE REVIEW

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
Looks good to me. Thanks for tackling this.

**Review by yanavlasov — APPROVED** [review]

File: source/common/http/filter_manager.cc [hunk 1]
```diff
@@ -874,8 +878,17 @@ void DownstreamFilterManager::sendLocalReply(
   if (!filter_manager_callbacks_.responseHeaders().has_value() &&
       !filter_manager_callbacks_.informationalHeaders().has_value()) {
     // If the response has not started at all, send the response through the filter chain.
-    sendLocalReplyViaFilterChain(is_grpc_request, code, body, modify_headers, is_head_request,
-                                 grpc_status, details);
+
+    // We only prepare a local reply to execute later if we're actively
+    // invoking filters to avoid re-entrant in filters.
+    if (avoid_reentrant_filter_invocation_during_local_reply_ &&
+        state_.decoder_filter_chain_aborted_) {
```
Comment (snowp) [inline source/common/http/filter_manager.cc:885]: Is `decoder_filter_chain_aborted_` a way to check whether we are actively invoking filters? Not immediately obvious why we are checking this from the comment

File: source/common/http/filter_manager.cc [hunk 2]
```diff
@@ -897,6 +910,56 @@ void DownstreamFilterManager::sendLocalReply(
   }
 }
 
+void DownstreamFilterManager::prepareLocalReplyViaFilterChain(
+    bool is_grpc_request, Code code, absl::string_view body,
+    const std::function<void(ResponseHeaderMap& headers)>& modify_headers, bool is_head_request,
+    const absl::optional<Grpc::Status::GrpcStatus> grpc_status, absl::string_view details) {
+  ENVOY_STREAM_LOG(debug, "Preparing sending local reply with details {}", *this, details);
```
Comment (snowp) [inline source/common/http/filter_manager.cc:917]:
```suggestion
  ENVOY_STREAM_LOG(debug, "Preparing local reply with details {}", *this, details);
```

File: source/common/http/filter_manager.h [hunk 3]
```diff
@@ -744,6 +745,16 @@ class FilterManager : public ScopeTrackedObject,
                               const std::function<void(ResponseHeaderMap& headers)>& modify_headers,
                               const absl::optional<Grpc::Status::GrpcStatus> grpc_status,
                               absl::string_view details) PURE;
+  /**
+   * Executes any prepared, but not yet propagated, local reply.
```
Comment (snowp) [inline source/common/http/filter_manager.h:749]:
```suggestion
   * Executes a prepared, but not yet propagated, local reply.
```

`any` makes it sound like there could be multiple, but there will only ever be one

File: source/common/http/utility.h [hunk 4]
```diff
@@ -391,6 +406,17 @@ struct LocalReplyData {
 void sendLocalReply(const bool& is_reset, const EncodeFunctions& encode_functions,
                     const LocalReplyData& local_reply_data);
 
+/**
+ * Prepares a locally generate response modifying headers and rewriting the
+ * response.
+ */
```
Comment (snowp) [inline source/common/http/utility.h:412]: Maybe just `Prepares a locally generated response` /w parameter docs? As is it is a bit hard to grok

File: test/extensions/filters/http/ext_proc/streaming_integration_test.cc [hunk 5]
```diff
@@ -671,11 +671,23 @@ TEST_P(StreamingIntegrationTest, PostAndProcessBufferedRequestBodyTooBig) {
         stream->Write(response);
 
         ProcessingRequest header_resp;
-        if (stream->Read(&header_resp)) {
-          ASSERT_TRUE(header_resp.has_response_headers());
+        bool seen_response_headers = false;
+
+        // Reading from the stream, we might receive the response headers
+        // later on depending on when the local reply is executed.
```
Comment (snowp) [inline test/extensions/filters/http/ext_proc/streaming_integration_test.cc:677]: What makes the execution non-deterministic?

File: test/integration/filters/add_invalid_data_filter.cc [hunk 6]
```diff
@@ -19,17 +19,26 @@ class AddInvalidDataFilter : public Http::PassThroughFilter {
...
-    decoder_callbacks_->addDecodedData(body, false);
+    if (!added_invalid_encode_) {
+      Buffer::OwnedImpl body("body");
+      decoder_callbacks_->addDecodedData(body, false);
+    }
     return Http::FilterHeadersStatus::Continue;
   }
+
+private:
+  // Track whether the user requested us to invoke make an invalid call to addEncodedData
```
Comment (snowp) [inline test/integration/filters/add_invalid_data_filter.cc:38]: This seems off, rephrase?

**Review by snowp — CHANGES_REQUESTED** [review]
Makes sense to me, thanks! A few comments

**Review by KBaichoo — COMMENTED** [review]
Thank you for the review @snowp

**Comment by KBaichoo** [comment]
Will merge main since DCO seems to be stuck.

**Review by snowp — CHANGES_REQUESTED** [review]
One nit, otherwise lgtm

File: test/integration/filters/assert_non_reentrant_filter.cc [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                                       "AssertNonReentrantFilter local reply during decodeHeaders.",
+                                       nullptr, absl::nullopt, "");
+    is_active_ = false;
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap&, bool) override {
+    ASSERT(!is_active_);
+    is_active_ = true;
+    is_active_ = false;
```
[the remark this answers is not in the document]
Comment (KBaichoo) [reply test/integration/filters/assert_non_reentrant_filter.cc]: removed since we're not doing anything interesting in between

**Review by snowp — APPROVED** [review]
Thanks!

**Pull request merged**